content/posts/2023-11-20-tools-for-learning-llvm-tablegen.md

```diff
@@ -0,0 +1,433 @@
+
+We’ve aimed to give the same experience as other languages, so I will focus not
+on how to use a notebook, but instead on what we’ve been able to make with them.
+
+## TableGen Tutorial
+
+The first notebook is an
+[introduction to TableGen](https://github.com/llvm/llvm-project/blob/main/llvm/utils/TableGen/jupyter/tablegen_tutorial_part_1.ipynb).
+You can read it on GitHub or download for use in Jupyter.
+
+The great thing about this is that if something in that document doesn’t make
+sense to you, you can edit the examples and experiment. Maybe it glosses over a
+part you find really interesting. You can add a new code and see what happens.
+
+## How to Write a TableGen Backend
+
+This notebook is different. The language is Python and instead of focusing on
+TableGen source, it will show you how to write a TableGen backend.
+
+The 2021 LLVM dev talk
+[“How to write a TableGen backend”](https://www.youtube.com/watch?v=UP-LBRbvI_U)
+by Min-Yih Hsu is the basis for this. The
+[notebook](https://github.com/llvm/llvm-project/blob/main/llvm/utils/TableGen/jupyter/sql_query_backend.ipynb)
+is in fact a Python port of Min’s own
+[C++](https://github.com/mshockwave/SQLGen) implementation.
+
+It shows you how to take the JSON output of `llvm-tblgen` and process it with
+Python to create SQL queries.
+
+What’s unique here is we now have the same content in multiple media forms and
+multiple programming languages. Pick the ones that best suit you.
+
+Referring back to “There is TableGen and There are Things Built With TableGen”
+, the tutorial notebook is TableGen. The writing a backend notebook is “Things
+Built With TableGen”.
+
+## Limitations
+
+The major limitation of the notebooks is that we have no output filtering. This
+means if you do `include “llvm/Target/Target.td"` you will get about 320,000
+lines of output (before you have added any of your own code). This is more than
+a default notebook will accept from a kernel and when I removed that limit, the
+browser tab crashed.
+
+It’s not a problem in most cases and the possible solutions have big trade offs
+so we are not rushing into fixing it. If it does affect you, please post your
+feedback on the [tracking issue](https://github.com/llvm/llvm-project/issues/72856).
+
+# TableGen Language Server
+
+The MLIR project has implemented a server for the
+[Language Server Protocol (LSP)](https://microsoft.github.io/language-server-protocol/).
+Which supports TableGen and
+[2 other languages](https://mlir.llvm.org/docs/Tools/MLIRLSP/) used within MLIR.
+
+The language server protocol is an editor agnostic means of providing
+information to an editor about the structure of a language and project. For
+instance, where are the include files, where is the definition of this type? If
+you have used an editor like Visual Studio Code, you’ve probably used this
+without knowing it. “Go To Definition” is the classic use case.
+
+It means you can open a project, go to the code you want to change and jump
```

Comment:
What does "It" refer to in this sentence? Do you mean the the TableGen Language Server? If so - suggest rewrite as follows:

The Language Server Protocol allows you to open a project, go to the code you want to change, and jump...